From 0bd2d50db651b6f59744e3445fa596bc298dd57e Mon Sep 17 00:00:00 2001
From: Pavel Tikhomirov <ptikhomirov@virtuozzo.com>
Date: Mon, 29 Nov 2021 12:54:51 +0300
Subject: [PATCH 164/245] mount: add mnt_bind_pick helper to pick the desired
 bind

Adding different pick functions we would be able to search different
things like mounted bind with wider root, or external bind, or external
bind with same sharing group and so on and so forth.

Signed-off-by: Pavel Tikhomirov <ptikhomirov@virtuozzo.com>
---
 criu/include/mount.h |  2 ++
 criu/mount.c         | 21 +++++++++++++++++++++
 2 files changed, 23 insertions(+)

diff --git a/criu/include/mount.h b/criu/include/mount.h
index 03959e1b0..c782d84bb 100644
--- a/criu/include/mount.h
+++ b/criu/include/mount.h
@@ -173,4 +173,6 @@ extern int remount_readonly_mounts(void);
 extern int try_remount_writable(struct mount_info *mi, bool ns);
 extern bool mnt_is_overmounted(struct mount_info *mi);
 
+extern struct mount_info *mnt_bind_pick(struct mount_info *mi,
+					bool (*pick)(struct mount_info *mi, struct mount_info *bind));
 #endif /* __CR_MOUNT_H__ */
diff --git a/criu/mount.c b/criu/mount.c
index de6722e69..fcdf06d50 100644
--- a/criu/mount.c
+++ b/criu/mount.c
@@ -934,6 +934,27 @@ static void search_bindmounts(void)
 		__search_bindmounts(mi);
 }
 
+struct mount_info *mnt_bind_pick(struct mount_info *mi, bool (*pick)(struct mount_info *mi, struct mount_info *bind))
+{
+	struct mount_info *bind;
+
+	BUG_ON(!mi);
+
+	if (pick(mi, mi))
+		return mi;
+
+	/*
+	 * Shouldn't use mnt_bind list before it was populated in search_bindmounts
+	 */
+	BUG_ON(!mi->mnt_bind_is_populated);
+
+	list_for_each_entry(bind, &mi->mnt_bind, mnt_bind)
+		if (pick(mi, bind))
+			return bind;
+
+	return NULL;
+}
+
 static int resolve_shared_mounts(struct mount_info *info, int root_master_id)
 {
 	struct mount_info *m, *t;
-- 
2.35.1

